### ISSUE TITLE
How to develop plugins?

### ISSUE THREAD

**Issue by Xerus2000** [sghpjuikit/player#44]
The Readme contains nothing about it. There definitely need to be instructions on how to code plugins.

**Comment by sghpjuikit** [comment]
Plugins were never moved out of their experimental status, due to missing their compilation framework, due to widget's framework not stabilizing. It's now on my todo list.

1 Plugin is a piece of functionality that can be started and stopped and normally, it would somehow manipulate the application (e.g. add new shortcuts). Each plugin also has its own user directory (user/plugins/pluginName) for resources and outputs.

2 See App.kt, at the bottom there is a short init plugin section where plugins are manually installed. Installing a plugin means to register an instance of a class implementing a Plugin (preferably PluginBase) using a App.pluginManager. As you can see the discovery and compilation mechanism is missing and as such the plugin must be provided and compiled with the application.

The plan (of my todo) is to make plugins work exactly the same as widgets. Each plugin will reside in its own directory somewhere under /plugins and be discovered and compiled in runtime by the app, using optional dependencies.

I'm open to suggestions at how this should work, although I'm fairly happy with this design.

Please, close the issue once the answer is satisfactory or keep asking.

**Comment by Xerus2000** [comment]
- what's the purpose of Screenrotator then? Like, it wouldn't even be able to be compiled with the app since it seems to be Windows binary stuff
- I don't know if that harsh distinction between widgets and plugins is neccessary, considering they work largely the same. I'm wondering if there will be any difference at startup time in the application between widget and plugin, or if it will only be a distinction based on what it contains
- a short readme note would still be helpful, denoting that they are experimental and not fully implemented

**Comment by sghpjuikit** [comment]
- ScreenRotator is a plugin that allows (on windows, using native program) to rotate screen using shortcuts (with multi--monitor setup support). I was tired of this feature missing in OS and some drivers so I made this little utility wrapper and made it into a plugin here for convenience. The plugin compiles fine on any platform, it simply will not work on non-Windows OS. I have not implemented an OS check as I'd like a more robust global reporting mechanism for widgets/plugins that they require a specific OS and such.
- the distinction is necessary. There is quite a gap, in life cycle, purpose, the way user should think and interact with the. Plugin always has only single instance. To produce layout, one needs to use Components - Containers to create the layout structure and Widgets to occupy the layout position. Widget is a Component, has graphics, uses a Controller (which produces behavior) and is not executed - it is merely loaded, used by the user and closed (through graphics). Plugin has no graphics to maintain its behavior (it only has settings), developer directly codes the Plugin class and the life cycle is different - plugin can be started and stopped, just like a Service, but Services are for the application and use registry/locator pattern to be discovered and accessible by application code, while plugins are anonymous and only meant for the user. In that sense Plugin joins the concept of Widget and Service.

In an OS analogy, Container is a window decoration, Widget is an application, Plugin is a service, Service is an internal service and Core is a shell. I will include a documentation about these 5 concepts in an application design/architecture section.

**Comment by sghpjuikit** [comment]
- `Plugin.isSupported` implemented.
- as advised Service & Plugin were merged. Although the distinction was supposed to be more pronounced, I ended up not exploiting that in the API*, the fact that Plugins are Services with no API is not justifiable for the duplication. All Service stuff is now Plugin stuff.

* which was:
- monadic Service container for safe Service instance control, but Plugins can leverage that too
- services were not supposed to be externally provided or recompilable, but this can be achieved by simply bundling them in the jar
